airbyte-integrations/connectors/source-s3/source_s3/v4/unstructured_parser.py

```diff
@@ -0,0 +1,94 @@
+
+from airbyte_cdk.sources.file_based.config.file_based_stream_config import FileBasedStreamConfig
+from airbyte_cdk.sources.file_based.file_based_stream_reader import AbstractFileBasedStreamReader, FileReadMode
+from airbyte_cdk.sources.file_based.file_types.file_type_parser import FileTypeParser
+from airbyte_cdk.sources.file_based.remote_file import RemoteFile
+from airbyte_cdk.sources.file_based.schema_helpers import SchemaType
+from source_s3.v4.config import S3FileBasedStreamConfig
+
+
+class UnstructuredParser(FileTypeParser):
+
+    MAX_SIZE_PER_CHUNK = 4_000_000
+
+    async def infer_schema(
+        self,
+        config: S3FileBasedStreamConfig,
+        file: RemoteFile,
+        stream_reader: AbstractFileBasedStreamReader,
+        logger: logging.Logger,
+    ) -> SchemaType:
+        return {
+            "content": {"type": "string"},
+            "chunk_number": {"type": "integer"},
+            "no_of_chunks": {"type": "integer"},
+            "id": {"type": "string"},
+        }
+
+    def parse_records(
+        self,
+        config: FileBasedStreamConfig,
+        file: RemoteFile,
+        stream_reader: AbstractFileBasedStreamReader,
+        logger: logging.Logger,
+        discovered_schema: Optional[Mapping[str, SchemaType]],
+    ) -> Iterable[Dict[str, Any]]:
+        with stream_reader.open_file(file, self.file_read_mode, None, logger) as fp:
+            markdown = self.read_file(fp, logger)
+            if not markdown:
+                return []
+            chunks = [markdown[i : i + self.MAX_SIZE_PER_CHUNK] for i in range(0, len(markdown), self.MAX_SIZE_PER_CHUNK)]
+            yield from [
+                {
+                    "content": chunk,
+                    "chunk_number": i,
+                    "no_of_chunks": len(chunks),
+                    "id": f"{file.uri}_{i}",
+                }
+                for i, chunk in enumerate(chunks)
+            ]
+
+    def read_file(self, file: IOBase, logger: logging.Logger) -> Optional[str]:
+        from unstructured.file_utils.filetype import FileType, detect_filetype
+        from unstructured.partition.auto import partition
+        from unstructured.partition.md import optional_decode
+
+        # set name to none, otherwise unstructured will try to get the modified date from the local file system
+        file_name = file.name
+        file.name = None
+        filetype = detect_filetype(
+            file=file,
+            file_filename=file_name,
+        )
+        if filetype == FileType.MD:
+            return optional_decode(file.read())
+        if filetype not in [FileType.PDF, FileType.DOCX]:
+            logger.warn(f"Skipping {file_name}, unsupported file type {str(filetype)}")
+            return None
+        elements = partition(file=file, metadata_filename=file_name)
+        return self._render_markdown(elements)
+
+    def _render_markdown(self, elements: List[Any]) -> str:
```

Comment:
I see that we need to keep 2 copies of the file in memory since `unstructured`'s `partition` function [returns a list](https://github.com/Unstructured-IO/unstructured/blob/8ab40c20c143ca3f809c830f3c0214a35939fbe4/unstructured/partition/pdf.py#L448), but when we call `self._render_markdown` it feels like we'll actually temporarily have up to 3 copies. I think we can do better than that by modifying `self._render_markdown` -

```suggestion
    def _render_markdown(self, elements: List[Any]) -> str:
        for el in elements:
            yield self._convert_to_markdown(el)
            yield "\n\n"
```
We'd then have to collect all the strings together in `parse_records` and strip the final `"\n\n"`, which isn't the prettiest thing in the world, but if this seems worth doing maybe we can find a cleaner way.

I'm also wondering, since both PDF and DOCX documents have a concept of pages, whether there are any libraries that would be able to break on the page boundary. In that case we could parse one page at a time, which may even avoid downloading the entire file at once depending on its implementation. Obviously that could interfere with the quality of `unstructured`'s parsing, but it may be something to think about if we're feeling like there are memory issues.

Reply:
I replaced this with a generator comprehension which shouldn't have the problem.

> I'm also wondering, since both PDF and DOCX documents have a concept of pages, whether there are any libraries that would be able to break on the page boundary

That's an interesting idea, I'm not sure whether it's possible or not to stream through a PDF or DOCX without keeping the full thing in memory. I'm also not sure how common this use case is (multi GB pdfs seem possible, but rare). IMHO we should keep it simple for the first version and rely on the lib that's making things simpler for us and break out of this abstraction if we get user feedback that it's an important feature.